### PR TITLE
feat: Aurora DSQL MCP Server

### PR DESCRIPTION
**Pull request by silver83**
## Summary
Add a DSQL MCP Server, supports natural-language interactions with a DSQL server.

### User experience

> See README.md changes

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [X] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [X] I have performed a self-review of this change
* [X] Changes have been tested
* [X] Changes are documented

Is this a breaking change? No

**RFC issue number**: https://github.com/awslabs/mcp/issues/346

Checklist:

* [N/A] Migration process documented
* [N/A] Implement warnings (if it can live side by side)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

### CODE REVIEW

**Review by scottschreckengaust — CHANGES_REQUESTED** [review]
Thank you for the contribution. Will you complete or associate an RFC.

File: README.md [hunk 1]
```diff
@@ -255,6 +256,14 @@ A server for generating, validating, and managing synthetic data.
 
 [Learn more](src/syntheticdata-mcp-server/README.md) | [Documentation](https://awslabs.github.io/mcp/servers/syntheticdata-mcp-server/)
 
+### Aurora DSQL MCP Server
+
+[![PyPI version](https://img.shields.io/pypi/v/awslabs.aurora-dsql-mcp-server.svg)](https://pypi.org/project/awslabs.git-repo-aurora-dsql-mcp-server/)
+
+An AWS Labs Model Context Protocol (MCP) server for Aurora Postgres
```
Comment (marcbowes) [inline README.md]: DSQL

File: docs/index.md [hunk 2]
```diff
@@ -92,6 +92,17 @@ The AWS Lambda MCP Server enables AI assistants to select and run AWS Lambda fun
...
 
+### Aurora DSQL MCP Server
+
+An AWS Labs Model Context Protocol (MCP) server for Aurora DSQL
+
+**Features:**
+
+- Execute queries against your DSQL Cluster
+
+Instructions: TODO
```
Comment (marcbowes) [inline docs/index.md]: Flagging

File: src/aurora-dsql-mcp-server/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    "awslabs.aurora-dsql-mcp-server": {
+      "command": "uvx",
+      "args": [
+         "awslabs.aurora-dsql-mcp-server@latest",
+         "--cluster_endpoint",
+         "[your dsql cluster endpoint]",
+         "--region",
+         "[your dsql cluster region, e.g. us-east-1]",
+         "--database_user",
+         "[your dsql username]"
```
Comment (marcbowes) [inline src/aurora-dsql-mcp-server/README.md:36]: I'd recommend adding `--profile` as an arg here.

File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+
+# Global variables
+cluster_endpoint = None
+database_user = None
+region = None
+read_only = False
+
+mcp = FastMCP(
+    'Aurora DSQL MCP server. This is the starting point for all solutions created',
```
Comment (marcbowes) [inline src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py]: Needs an update

File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    dependencies=[
+        'loguru',
+    ],
+)
+
+
+@mcp.tool(name='query', description='Run a read-only SQL query')
```
Comment (marcbowes) [inline src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py]: What about readonly_query?

'query' can sometimes mean updates too

File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+    parser.add_argument('--port', type=int, default=8888, help='Port to run the server on')
+    parser.add_argument(
+        '--cluster_endpoint', required=True, help='Endpoint for your Aurora DSQL cluster'
+    )
+    parser.add_argument('--database_user', required=True, help='Database username')
+    parser.add_argument(
+        '--region',
+        required=True,
+        default='us-west-2',
+        help='AWS region for Aurora DSQL Cluster (default: us-west-2)',
```
Comment (marcbowes) [inline src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py]: Would recommend not having a default and letting boto default based on AWS_REGION

**Comment by silver83** [comment]
> Thank you for the contribution. Will you complete or associate an RFC.

@scottschreckengaust thank you, created https://github.com/awslabs/mcp/issues/346

**Review by marcbowes — COMMENTED** [review]
Looks good, thanks!

File: src/aurora-dsql-mcp-server/README.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+## Features
+
+- Converting human-readable questions and commands into structured Postgres-compatible SQL queries and executing them against the configured Aurora DSQL database.
+- Read-only mode via --read-only
+
+## Prerequisites
+
+1. Install `uv` from [Astral](https://docs.astral.sh/uv/getting-started/installation/) or the [GitHub README](https://github.com/astral-sh/uv#installation)
+2. Install Python using `uv python install 3.10`
```
Comment (kennthhz) [inline src/aurora-dsql-mcp-server/README.md:13]: should also support Docker option to be consistent with other MCP server (especially Aurora)

File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+    """
+    logger.info(f'query: {sql}')
+
+    if not sql:
+        await ctx.error(ERROR_EMPTY_SQL_PASSED_TO_READONLY_QUERY)
+        raise ValueError(ERROR_EMPTY_SQL_PASSED_TO_READONLY_QUERY)
+
+    conn = await create_connection(ctx)
+
+    await execute_query(ctx, conn, 'BEGIN TRANSACTION READ ONLY')
```
Comment (kennthhz) [inline src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py]: nit: why not put this const string in consts.py

File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+    try:
+        rows = await execute_query(ctx, conn, sql)
+        await execute_query(ctx, conn, 'COMMIT')
+        return rows
+    except Exception as e:
```
Comment (kennthhz) [inline src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py:151]: I'd suggest separate the expected and unexpected exceptions

Comment (kennthhz) [reply src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py:151]: Should we raise exception to FastMCP? I thought the pattern is await ctx.error() and return error response. At least that is what other existing MCP servers implemented

File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+        await execute_query(ctx, conn, 'ROLLBACK')
+        raise e
+    finally:
+        await conn.close()
+
+
+@mcp.tool(name='schema', description='Get the schema of the given table')
+async def schema(
```
Comment (kennthhz) [inline src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py]: suggest naming the function as get_schema

**Review by kennthhz — CHANGES_REQUESTED** [review]
where is pyproject.toml? I think you need it along with Dockerfile

**Comment by awslabs-mcp** [comment]
## General Assessment

I've reviewed PR #328 which adds the Aurora DSQL MCP Server. This server enables natural language interaction with Aurora DSQL databases, providing tools to execute SQL queries, run transactions, and retrieve schema information.

The implementation follows the MCP server design guidelines, with proper error handling, authentication, and security measures like the default read-only mode. The code is well-structured with a clear separation between the server implementation, constants, and tests.

### Strengths
- Well-documented code with clear function docstrings
- Proper error handling with specific error messages
- Security-focused design with default read-only mode
- Connection validation at startup
- Transaction management with proper BEGIN/COMMIT/ROLLBACK flow
- Comprehensive test coverage (though Codecov reports 18 missing lines)

### Areas for Improvement
- Line 168: Consider providing more context in the error message from the schema tool
- Line 226-228: Consider using a more explicit condition to check for results
- Constant  is defined but not used in the code
- The PyPI badge in the main README.md points to 'awslabs.git-repo-aurora-dsql-mcp-server' which appears to be incorrect

Overall, this is a well-implemented MCP server that follows the project's design guidelines and provides valuable functionality for interacting with Aurora DSQL databases.

Run ID: 15076773303

**Comment by awslabs-mcp** [comment]
### File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/consts.py

 is defined but doesn't appear to be used in the server implementation. If there's no 'explain' tool in this version, consider removing this constant or adding a TODO comment.

**Comment by awslabs-mcp** [comment]
### File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py

Line 168: Consider adding more context in the exception being raised, similar to how you handle other errors. Currently, it just re-raises the exception without adding any context that would help the client understand what went wrong.

**Comment by awslabs-mcp** [comment]
### File: README.md

The PyPI badge URL appears to be incorrect. It currently points to 'awslabs.git-repo-aurora-dsql-mcp-server' but should likely be 'awslabs.aurora-dsql-mcp-server' based on the package name.

**Review by awslabs-mcp — APPROVED** [review]
## Final Assessment

I've reviewed PR #328 which adds the Aurora DSQL MCP Server and found it to be well-implemented according to project standards. The code quality is good with appropriate error handling, security measures, and documentation.

### Minor Issues Identified:
- Unused error constant 
- Missing error context in schema tool exception handling
- Incorrect PyPI badge URL in README.md

These are minor issues that could be addressed in future PRs and don't impact the functionality of the server.

The implementation follows best practices from the project's design guidelines and provides valuable functionality for interacting with Aurora DSQL databases through natural language queries.

I approve this PR as the changes meet quality standards and the server will be a valuable addition to the MCP ecosystem.

Run ID: 15076773303

**Review by awslabs-mcp — APPROVED** [review]
## Final Assessment

I've reviewed PR #328 which adds the Aurora DSQL MCP Server and found it to be well-implemented according to project standards. The code quality is good with appropriate error handling, security measures, and documentation.

### Minor Issues Identified:
- Unused error constant 'ERROR_EMPTY_SQL_PASSED_TO_EXPLAIN'
- Missing error context in schema tool exception handling
- Incorrect PyPI badge URL in README.md

These are minor issues that could be addressed in future PRs and don't impact the functionality of the server.

The implementation follows best practices from the project's design guidelines and provides valuable functionality for interacting with Aurora DSQL databases through natural language queries.

I approve this PR as the changes meet quality standards and the server will be a valuable addition to the MCP ecosystem.

Run ID: 15076773303

File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+    ### get_schema
+    Returns the schema of a table.
+    """,
+    dependencies=[
+        'loguru',
+    ],
+)
+
+
+@mcp.tool(name='readonly_query', description='Run a read-only SQL query')
```
Comment (marcbowes) [inline src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py]:
```suggestion
@mcp.tool(name='readonly_query', description='''Run a read-only SQL query against the configured Aurora DSQL cluster.

Aurora DSQL is distributed SQL database with Postgres compatibility. The following table
summarizes `SELECT` functionality that is expected to work. Items not in this table may
also be supported, as this is a point in time snapshot.

| Primary clause                  | Supported clauses     |
|---------------------------------|-----------------------|
| FROM                            |                       |
| GROUP BY                        | ALL, DISTINCT         |
| ORDER BY                        | ASC, DESC, NULLS      |
| LIMIT                           |                       |
| DISTINCT                        |                       |
| HAVING                          |                       |
| USING                           |                       |
| WITH (common table expressions) |                       |
| INNER JOIN                      | ON                    |
| OUTER JOIN                      | LEFT, RIGHT, FULL, ON |
| CROSS JOIN                      | ON                    |
| UNION                           | ALL                   |
| INTERSECT                       | ALL                   |
| EXCEPT                          | ALL                   |
| OVER                            | RANK (), PARTITION BY |
| FOR UPDATE                      |                       |
''')
```

File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+            except Exception as e:
+                logger.error(f'{ERROR_ROLLBACK_TRANSACTION}: {str(e)}')
+            await conn.close()
+
+    except Exception as e:
+        await ctx.error(f'{ERROR_READONLY_QUERY}: {str(e)}')
+        raise Exception(f'{ERROR_READONLY_QUERY}: {str(e)}')
+
+
+@mcp.tool(name='transact', description='Write or modify data using SQL, in a transaction')
```
Comment (marcbowes) [inline src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py]:
```suggestion
@mcp.tool(name='transact', description='''Write or modify data using SQL, in a transaction against the configured Aurora DSQL cluster.

Aurora DSQL is a distributed SQL database with Postgres compatibility. This tool will automatically
insert `BEGIN` and `COMMIT` statements; you only need to provide the statements to run
within the transaction scope.

In addition to the `SELECT` functionality described on the `readonly_query` tool, DSQL supports
common DDL statements such as `CREATE TABLE`. Note that it is a best practice to use UUIDs
for new tables in DSQL as this will spread your workload out over as many nodes as possible.

Some DDL commands are async (like `CREATE INDEX ASYNC`), and return a job id. Jobs can
be viewed by running `SELECT * FROM sys.jobs`.
```

File: README.md [hunk 13]
```diff
@@ -343,6 +344,14 @@ A server for generating, validating, and managing synthetic data.
...
 [Learn more](src/syntheticdata-mcp-server/README.md) | [Documentation](https://awslabs.github.io/mcp/servers/syntheticdata-mcp-server/)
 
+### Aurora DSQL MCP Server
+
+[![PyPI version](https://img.shields.io/pypi/v/awslabs.aurora-dsql-mcp-server.svg)](https://pypi.org/project/awslabs.aurora-dsql-mcp-server/)
+
+An AWS Labs Model Context Protocol (MCP) server for Aurora DSQL
+
+- Converting human-readable questions and commands into structured Postgres-compatible SQL queries and executing them against the configured Aurora DSQL database.
+
```
Comment (krokoko) [inline README.md:463]: Could you please add "[Learn more](link) | [Documentation](link)" as other servers have in this readme ?

File: src/aurora-dsql-mcp-server/README.md (hunk 7 above)
Comment (krokoko) [reply src/aurora-dsql-mcp-server/README.md:13]: +1

**Review by marcbowes — COMMENTED** [review]
🚀

**Review by krokoko — APPROVED** [review]

**Review by awslabs-mcp — COMMENTED** [review]
### File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py

In the get_schema function (lines 229-253), when an exception is raised at line 252, it would be better to include additional context in the error message, similar to how other error handlers in this file are implemented. Instead of just passing the original error, consider adding more specific details about what failed.

In the execute_query function (lines 301-320), the condition check at line 311 with  could be more explicit about what this check is testing for (no results vs. DDL statement that doesn't return rows).

**Review by awslabs-mcp — COMMENTED** [review]
### File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/server.py

In the get_schema function (lines 229-253), when an exception is raised at line 252, it would be better to include additional context in the error message, similar to how other error handlers in this file are implemented. Instead of just passing the original error, consider adding more specific details about what failed.

In the execute_query function (lines 301-320), the condition check at line 311 with 'if cur.rownumber is None:' could be more explicit about what this check is testing for (no results vs. DDL statement that doesn't return rows).

**Review by awslabs-mcp — COMMENTED** [review]
### File: src/aurora-dsql-mcp-server/awslabs/aurora_dsql_mcp_server/consts.py

The consts.py file is missing any constant related to 'EXPLAIN' functionality, although looking at the PR comments there might have been plans to add an 'explain' tool at some point. If there are no immediate plans for this feature, that's fine, but consider adding a TODO comment if this functionality is intended for a future update.

**Review by awslabs-mcp — COMMENTED** [review]
### File: src/aurora-dsql-mcp-server/README.md

The PyPI badge URL in the README.md appears to be incorrect. The badge should point to 'awslabs.aurora-dsql-mcp-server' but the current URL structure suggests it's referencing 'awslabs.git-repo-aurora-dsql-mcp-server'. Please update the badge URL to ensure it points to the correct PyPI package.

**Review by awslabs-mcp — APPROVED** [review]
## Final Assessment\n\nI've conducted a comprehensive review of PR #328 adding the Aurora DSQL MCP Server. This server enables natural language interaction with Amazon Aurora DSQL databases through tools for executing queries, running transactions, and retrieving schema information.\n\n### Strengths:\n\n- Well-structured codebase following the MCP server design guidelines\n- Strong security measures including default read-only mode requiring explicit opt-in for write operations\n- Comprehensive error handling with contextual error messages\n- Good test coverage with unit tests for all main functions including edge cases\n- Clean separation of concerns between connection management, query execution, and transaction handling\n- Clear documentation in README.md explaining setup options and limitations\n\n### Areas for Improvement:\n\n- The get_schema function's error handling could provide more context (lines 229-253)\n- The execute_query function's condition check for empty results could be more explicit (lines 301-320)\n- The PyPI badge URL in README.md references an incorrect package name\n- There's a potential unused constant ERROR_EMPTY_SQL_PASSED_TO_EXPLAIN in imports\n\nThese are minor issues that don't impact the core functionality and could be addressed in future PRs.\n\nThe implementation aligns well with the codebase's conventions, provides appropriate security measures (especially the default read-only mode), and adds valuable functionality for working with Aurora DSQL databases. The code is well-tested and follows the established patterns for MCP servers.\n\nI approve this PR as it meets the project's quality standards and will be a valuable addition to the MCP ecosystem.\n\nRun ID: 15164307261
